core/src/main/java/io/grpc/internal/InternalSubchannel.java

```diff
@@ -286,6 +286,28 @@ void resetConnectBackoff() {
     }
   }
 
+  void shutdownTransports(Status reason) {
```

Comment:
Using this new method may break the contract of `ManagedClientTransport` interface:
```java
  /**
   * ... ...  This method may only be called once.
   */
  void shutdown(Status reason);
```
Why not just use the existing `InternalSubchannel.shutdown()` method?

Reply:
The current API assumes that `InternalSubchannel.shutdown()` only occurs when the LB has discarded the subchannel or the managed channel itself is shutting down. See the Javadoc on `io.grpc.LoadBalancer#handleSubchannelState`. `InternalSubchannel.shutdown()` is purposefully ignored by the LB, so just invoking this method means that outgoing RPCs will still be sent to the shutdown subchannel.

We can change this behavior in the LB/somewhere else to allow just shutting down the `InternalSubchannel` directly and recovering properly, but this is a more extensive change than just shutting down the transports as done here. Thoughts?

Reply:
One option may be: keep your code and add 
```java
activeTransport = null;
pendingTransport = null;
addressIndex = 0;
```
 
after
```java
savedActiveTransport = activeTransport;
savedPendingTransport = pendingTransport;
```

Another option: call `InternalSubchannel.shutdown()` and  change the behavior of LB so that is does not ignore it completely (LB ignored it because of the assumptions you mentioned which were the only two cases we knew at that time, this is never a hard requirement, now that we have a new case, maybe it should be changed.)

Reply:
Went with option 1. When the `InternalSubchannel` itself is shutting down, we can ignore the resulting `TransportListener.transportShutdown` callback, but here we still want to handle the callback (as if the transport itself caused the shutdown), so saved an additional reference to the shutdown channels for use in the callback.